File: src/binding.ts
import type { Program } from "./ast.js";
import { error, type Diagnostic } from "./diagnostics.js";

export interface BindingEntrySpec {
  readonly id: string;
  readonly auditName: string;
  readonly programName: string;
}

export interface BindingObjectSpec extends BindingEntrySpec {
  readonly fields: readonly BindingEntrySpec[];
}

export interface BindingFlowSpec extends BindingEntrySpec {
  readonly inputs: readonly BindingEntrySpec[];
  readonly outputs: readonly BindingEntrySpec[];
}

export interface BindingSpec {
  readonly version: 1;
  readonly objects: readonly BindingObjectSpec[];
  readonly flows: readonly BindingFlowSpec[];
}

export interface ResolvedObjectBinding {
  readonly id: string;
  readonly auditName: string;
  readonly programName: string;
  readonly fields: ReadonlyMap<string, BindingEntrySpec>;
}

export interface ResolvedFlowBinding {
  readonly id: string;
  readonly auditName: string;
  readonly programName: string;
  readonly inputs: ReadonlyMap<string, BindingEntrySpec>;
  readonly outputs: ReadonlyMap<string, BindingEntrySpec>;
}

export interface ResolvedBinding {
  readonly objects: ReadonlyMap<string, ResolvedObjectBinding>;
  readonly flows: ReadonlyMap<string, ResolvedFlowBinding>;
}

export interface BindingResult {
  readonly binding: ResolvedBinding | null;
  readonly diagnostics: readonly Diagnostic[];
}

const LOCATION = { line: 1, column: 1 } as const;

export function parseBinding(source: string): { readonly spec: BindingSpec | null; readonly diagnostics: readonly Diagnostic[] } {
  let value: unknown;
  try {
    value = JSON.parse(source) as unknown;
  } catch (cause) {
    return { spec: null, diagnostics: [error(`绑定文件不是有效 JSON：${cause instanceof Error ? cause.message : String(cause)}`, LOCATION)] };
  }

  const diagnostics: Diagnostic[] = [];
  if (!isRecord(value) || value.version !== 1) {
    diagnostics.push(error("绑定文件必须包含 version: 1", LOCATION));
    return { spec: null, diagnostics };
  }
  const objects = readObjectSpecs(value.objects, diagnostics);
  const flows = readFlowSpecs(value.flows, diagnostics);
  if (!objects || !flows) return { spec: null, diagnostics };
  return { spec: { version: 1, objects, flows }, diagnostics };
}

export function resolveBinding(program: Program, spec?: BindingSpec): BindingResult {
  const diagnostics: Diagnostic[] = [];
  const actual = spec ?? createIdentityBinding(program);
  const objects = new Map<string, ResolvedObjectBinding>();
  const flows = new Map<string, ResolvedFlowBinding>();
  const ids = new Set<string>();
  const objectNames = new Set<string>();
  const flowNames = new Set<string>();

  for (const object of program.objects) {
    const entry = actual.objects.find((candidate) => candidate.auditName === object.name);
    if (!entry) {
      diagnostics.push(error(`对象缺少绑定：${object.name}`, object.loc));
      continue;
    }
    validateTopLevelEntry(entry, `对象 ${object.name}`, ids, objectNames, diagnostics);
    const fields = resolveMembers(object.fields.map((field) => field.name), entry.fields, `对象 ${object.name} 的字段`, ids, diagnostics);
    objects.set(object.name, { ...entry, fields });
  }
  for (const entry of actual.objects) {
    if (!program.objects.some((object) => object.name === entry.auditName)) diagnostics.push(error(`绑定引用了未声明的对象：${entry.auditName}`, LOCATION));
  }

  for (const flow of program.flows) {
    const entry = actual.flows.find((candidate) => candidate.auditName === flow.name);
    if (!entry) {
      diagnostics.push(error(`流程缺少绑定：${flow.name}`, flow.loc));
      continue;
    }
    validateTopLevelEntry(entry, `流程 ${flow.name}`, ids, flowNames, diagnostics);
    const inputs = resolveMembers(flow.inputs.map((input) => input.name), entry.inputs, `流程 ${flow.name} 的输入`, ids, diagnostics);
    const outputs = resolveMembers(flow.outputs.map((output) => output.name), entry.outputs, `流程 ${flow.name} 的输出`, ids, diagnostics);
    flows.set(flow.name, { ...entry, inputs, outputs });
  }
  for (const entry of actual.flows) {
    if (!program.flows.some((flow) => flow.name === entry.auditName)) diagnostics.push(error(`绑定引用了未声明的流程：${entry.auditName}`, LOCATION));
  }

  return { binding: diagnostics.length === 0 ? { objects, flows } : null, diagnostics };
}

export function createIdentityBinding(program: Program): BindingSpec {
  return {
    version: 1,
    objects: program.objects.map((object, objectIndex) => ({
      id: `object_${objectIndex}`,
      auditName: object.name,
      programName: `Object_${objectIndex}`,
      fields: object.fields.map((field, fieldIndex) => ({
        id: `field_${objectIndex}_${fieldIndex}`,
        auditName: field.name,
        programName: field.name,
      })),
    })),
    flows: program.flows.map((flow, flowIndex) => ({
      id: `flow_${flowIndex}`,
      auditName: flow.name,
      programName: `flow_${flowIndex}`,
      inputs: flow.inputs.map((input, inputIndex) => ({
        id: `input_${flowIndex}_${inputIndex}`,
        auditName: input.name,
        programName: input.name,
      })),
      outputs: flow.outputs.map((output, outputIndex) => ({
        id: `output_${flowIndex}_${outputIndex}`,
        auditName: output.name,
        programName: output.name,
      })),
    })),
  };
}

export function bindingFingerprint(binding: ResolvedBinding): string {
  const canonical = {
    objects: [...binding.objects.entries()].sort(([left], [right]) => compareNames(left, right)).map(([name, object]) => ({
      name,
      id: object.id,
      programName: object.programName,
      fields: [...object.fields.entries()].sort(([left], [right]) => compareNames(left, right)).map(([fieldName, field]) => ({ name: fieldName, ...field })),
    })),
    flows: [...binding.flows.entries()].sort(([left], [right]) => compareNames(left, right)).map(([name, flow]) => ({
      name,
      id: flow.id,
      programName: flow.programName,
      inputs: [...flow.inputs.entries()].sort(([left], [right]) => compareNames(left, right)).map(([inputName, input]) => ({ name: inputName, ...input })),
      outputs: [...flow.outputs.entries()].sort(([left], [right]) => compareNames(left, right)).map(([outputName, output]) => ({ name: outputName, ...output })),
    })),
  };
  let hash = 2166136261;
  for (const character of JSON.stringify(canonical)) {
    hash ^= character.charCodeAt(0);
    hash = Math.imul(hash, 16777619) >>> 0;
  }
  return hash.toString(16).padStart(8, "0");
}

function compareNames(left: string, right: string): number {
  return left < right ? -1 : left > right ? 1 : 0;
}

function resolveMembers(names: readonly string[], entries: readonly BindingEntrySpec[], label: string, ids: Set<string>, diagnostics: Diagnostic[]): ReadonlyMap<string, BindingEntrySpec> {
  const expected = new Set(names);
  const resolved = new Map<string, BindingEntrySpec>();
  const programNames = new Set<string>();
  for (const entry of entries) {
    if (!expected.has(entry.auditName)) {
      diagnostics.push(error(`${label}绑定了未声明的名称：${entry.auditName}`, LOCATION));
      continue;
    }
    validateMemberEntry(entry, label, ids, programNames, diagnostics);
    if (resolved.has(entry.auditName)) diagnostics.push(error(`${label}绑定重复：${entry.auditName}`, LOCATION));
    resolved.set(entry.auditName, entry);
  }
  for (const name of names) if (!resolved.has(name)) diagnostics.push(error(`${label}缺少绑定：${name}`, LOCATION));
  return resolved;
}

function validateTopLevelEntry(entry: BindingEntrySpec, label: string, ids: Set<string>, programNames: Set<string>, diagnostics: Diagnostic[]): void {
  validateEntry(entry, label, ids, diagnostics);
  if (!isJavaScriptIdentifier(entry.programName)) diagnostics.push(error(`${label}的程序名称必须是合法标识符：${entry.programName}`, LOCATION));
  if (programNames.has(entry.programName)) diagnostics.push(error(`${label}的程序名称重复：${entry.programName}`, LOCATION));
  programNames.add(entry.programName);
}

function validateMemberEntry(entry: BindingEntrySpec, label: string, ids: Set<string>, programNames: Set<string>, diagnostics: Diagnostic[]): void {
  validateEntry(entry, label, ids, diagnostics);
  if (entry.programName.trim().length === 0) diagnostics.push(error(`${label}的程序名称不能为空`, LOCATION));
  if (programNames.has(entry.programName)) diagnostics.push(error(`${label}的程序名称重复：${entry.programName}`, LOCATION));
  programNames.add(entry.programName);
}

function validateEntry(entry: BindingEntrySpec, label: string, ids: Set<string>, diagnostics: Diagnostic[]): void {
  if (!/^[a-z][a-z0-9_-]*$/u.test(entry.id)) diagnostics.push(error(`${label}的内部身份必须是稳定的小写标识：${entry.id}`, LOCATION));
  if (ids.has(entry.id)) diagnostics.push(error(`绑定内部身份重复：${entry.id}`, LOCATION));
  ids.add(entry.id);
  if (entry.auditName.trim().length === 0) diagnostics.push(error(`${label}的审计名称不能为空`, LOCATION));
}

function readObjectSpecs(value: unknown, diagnostics: Diagnostic[]): BindingObjectSpec[] | null {
  if (!Array.isArray(value)) {
    diagnostics.push(error("绑定文件的 objects 必须是数组", LOCATION));
    return null;
  }
  const result: BindingObjectSpec[] = [];
  for (const [index, item] of value.entries()) {
    const entry = readEntry(item, `objects[${index}]`, diagnostics);
    if (!entry) continue;
    const fields = readEntries(isRecord(item) ? item.fields : undefined, `objects[${index}].fields`, diagnostics);
    if (fields) result.push({ ...entry, fields });
  }
  return result;
}

function readFlowSpecs(value: unknown, diagnostics: Diagnostic[]): BindingFlowSpec[] | null {
  if (!Array.isArray(value)) {
    diagnostics.push(error("绑定文件的 flows 必须是数组", LOCATION));
    return null;
  }
  const result: BindingFlowSpec[] = [];
  for (const [index, item] of value.entries()) {
    const entry = readEntry(item, `flows[${index}]`, diagnostics);
    if (!entry) continue;
    const inputs = readEntries(isRecord(item) ? item.inputs : undefined, `flows[${index}].inputs`, diagnostics);
    const outputs = readEntries(isRecord(item) ? item.outputs : undefined, `flows[${index}].outputs`, diagnostics);
    if (inputs && outputs) result.push({ ...entry, inputs, outputs });
  }
  return result;
}

function readEntries(value: unknown, label: string, diagnostics: Diagnostic[]): BindingEntrySpec[] | null {
  if (!Array.isArray(value)) {
    diagnostics.push(error(`${label} 必须是数组`, LOCATION));
    return null;
  }
  const result: BindingEntrySpec[] = [];
  for (const [index, item] of value.entries()) {
    const entry = readEntry(item, `${label}[${index}]`, diagnostics);
    if (entry) result.push(entry);
  }
  return result;
}

function readEntry(value: unknown, label: string, diagnostics: Diagnostic[]): BindingEntrySpec | null {
  if (!isRecord(value) || typeof value.id !== "string" || typeof value.auditName !== "string" || typeof value.programName !== "string") {
    diagnostics.push(error(`${label} 必须包含 id、auditName 和 programName`, LOCATION));
    return null;
  }
  return { id: value.id, auditName: value.auditName, programName: value.programName };
}

function isRecord(value: unknown): value is Record<string, unknown> {
  return Boolean(value && typeof value === "object" && !Array.isArray(value));
}

function isJavaScriptIdentifier(value: string): boolean {
  return /^[A-Za-z_$][A-Za-z0-9_$]*$/u.test(value) && !JAVA_SCRIPT_RESERVED_WORDS.has(value);
}

const JAVA_SCRIPT_RESERVED_WORDS = new Set([
  "break", "case", "catch", "class", "const", "continue", "debugger", "default", "delete", "do", "else", "export", "extends", "finally", "for", "function", "if", "import", "in", "instanceof", "let", "new", "return", "super", "switch", "this", "throw", "try", "typeof", "var", "void", "while", "with", "yield",
]);

File: internal/binding/binding.go
package binding

import (
	"encoding/json"
	"fmt"
	"regexp"

	"github.com/myczh-1/determinant/internal/ast"
	"github.com/myczh-1/determinant/internal/diagnostics"
)

type Entry struct {
	ID          string `json:"id"`
	AuditName   string `json:"auditName"`
	ProgramName string `json:"programName"`
}

type ObjectSpec struct {
	Entry
	Fields []Entry `json:"fields"`
}

type FlowSpec struct {
	Entry
	Inputs  []Entry `json:"inputs"`
	Outputs []Entry `json:"outputs"`
}

type Spec struct {
	Version int          `json:"version"`
	Objects []ObjectSpec `json:"objects"`
	Flows   []FlowSpec   `json:"flows"`
}

type Object struct {
	ID          string
	AuditName   string
	ProgramName string
	Fields      map[string]Entry
}

type Flow struct {
	ID          string
	AuditName   string
	ProgramName string
	Inputs      map[string]Entry
	Outputs     map[string]Entry
}

type Resolved struct {
	Objects map[string]Object
	Flows   map[string]Flow
}

var stableIDPattern = regexp.MustCompile(`^[a-z][a-z0-9_-]*$`)
var programNamePattern = regexp.MustCompile(`^[\p{L}_$][\p{L}\p{N}_$]*$`)

func Parse(data []byte, file string) (Spec, []diagnostics.Diagnostic) {
	var spec Spec
	if err := json.Unmarshal(data, &spec); err != nil {
		return Spec{}, []diagnostics.Diagnostic{{Severity: diagnostics.Error, Code: "AAL2200", Message: "binding file is not valid JSON: " + err.Error(), File: file, Line: 1, Column: 1}}
	}
	result := []diagnostics.Diagnostic{}
	if spec.Version != 1 {
		result = append(result, diagnostic(file, "AAL2201", "binding file must declare version: 1"))
	}
	if spec.Objects == nil {
		result = append(result, diagnostic(file, "AAL2202", "binding objects must be an array"))
	}
	if spec.Flows == nil {
		result = append(result, diagnostic(file, "AAL2203", "binding flows must be an array"))
	}
	ids := map[string]bool{}
	for _, object := range spec.Objects {
		validateEntry(object.Entry, "object "+object.AuditName, ids, &result, file)
		if object.Fields == nil {
			result = append(result, diagnostic(file, "AAL2204", "binding fields must be an array for object "+object.AuditName))
		}
		for _, field := range object.Fields {
			validateEntry(field, "field "+object.AuditName+"."+field.AuditName, ids, &result, file)
		}
		validateUniqueMemberNames(object.Fields, "field "+object.AuditName, &result, file)
	}
	for _, flow := range spec.Flows {
		validateEntry(flow.Entry, "flow "+flow.AuditName, ids, &result, file)
		for _, input := range flow.Inputs {
			validateEntry(input, "input "+flow.AuditName+"."+input.AuditName, ids, &result, file)
		}
		for _, output := range flow.Outputs {
			validateEntry(output, "output "+flow.AuditName+"."+output.AuditName, ids, &result, file)
		}
		validateUniqueMemberNames(flow.Inputs, "input "+flow.AuditName, &result, file)
		validateUniqueMemberNames(flow.Outputs, "output "+flow.AuditName, &result, file)
	}
	validateNamedEntries(spec.Objects, "object", &result, file)
	validateNamedFlows(spec.Flows, &result, file)
	return spec, result
}

func Resolve(program *ast.Program, spec *Spec) (*Resolved, []diagnostics.Diagnostic) {
	if program == nil {
		return nil, []diagnostics.Diagnostic{{Severity: diagnostics.Error, Code: "AAL2205", Message: "cannot resolve binding without a parsed program"}}
	}
	actual := spec
	if actual == nil {
		identity := Identity(program)
		actual = &identity
	}
	result := []diagnostics.Diagnostic{}
	resolved := &Resolved{Objects: map[string]Object{}, Flows: map[string]Flow{}}
	for _, object := range program.Objects {
		entry, ok := findObjectSpec(actual.Objects, object.Name)
		if !ok {
			result = append(result, at(object.Location, "AAL2206", "object is missing from binding: "+object.Name))
			continue
		}
		fields := map[string]Entry{}
		for _, field := range object.Fields {
			mapped, ok := findEntry(entry.Fields, field.Name)
			if !ok {
				result = append(result, at(field.Location, "AAL2207", "object field is missing from binding: "+object.Name+"."+field.Name))
				continue
			}
			fields[field.Name] = mapped
		}
		for _, mapped := range entry.Fields {
			if !hasObjectField(object, mapped.AuditName) {
				result = append(result, at(object.Location, "AAL2222", "binding references undeclared object field: "+object.Name+"."+mapped.AuditName))
			}
		}
		resolved.Objects[object.Name] = Object{ID: entry.ID, AuditName: entry.AuditName, ProgramName: entry.ProgramName, Fields: fields}
	}
	for _, entry := range actual.Objects {
		if _, ok := find(program.Objects, entry.AuditName); !ok {
			result = append(result, diagnostic("", "AAL2208", "binding references undeclared object: "+entry.AuditName))
		}
	}
	for _, flow := range program.Flows {
		entry, ok := findFlowSpec(actual.Flows, flow.Name)
		if !ok {
			result = append(result, at(flow.Location, "AAL2209", "flow is missing from binding: "+flow.Name))
			continue
		}
		inputs, outputs := map[string]Entry{}, map[string]Entry{}
		for _, input := range flow.Inputs {
			mapped, ok := findEntry(entry.Inputs, input.Name)
			if !ok {
				result = append(result, at(input.Location, "AAL2210", "flow input is missing from binding: "+flow.Name+"."+input.Name))
				continue
			}
			inputs[input.Name] = mapped
		}
		for _, output := range flow.Outputs {
			mapped, ok := findEntry(entry.Outputs, output.Name)
			if !ok {
				result = append(result, at(output.Location, "AAL2211", "flow output is missing from binding: "+flow.Name+"."+output.Name))
				continue
			}
			outputs[output.Name] = mapped
		}
		for _, mapped := range entry.Inputs {
			if _, ok := findFlowInput(flow, mapped.AuditName); !ok {
				result = append(result, at(flow.Location, "AAL2223", "binding references undeclared flow input: "+flow.Name+"."+mapped.AuditName))
			}
		}
		for _, mapped := range entry.Outputs {
			if _, ok := findFlowOutput(flow, mapped.AuditName); !ok {
				result = append(result, at(flow.Location, "AAL2224", "binding references undeclared flow output: "+flow.Name+"."+mapped.AuditName))
			}
		}
		resolved.Flows[flow.Name] = Flow{ID: entry.ID, AuditName: entry.AuditName, ProgramName: entry.ProgramName, Inputs: inputs, Outputs: outputs}
	}
	for _, entry := range actual.Flows {
		if !hasFlow(program, entry.AuditName) {
			result = append(result, diagnostic("", "AAL2212", "binding references undeclared flow: "+entry.AuditName))
		}
	}
	if len(result) > 0 {
		return nil, result
	}
	return resolved, result
}

func Identity(program *ast.Program) Spec {
	spec := Spec{Version: 1}
	for objectIndex, object := range program.Objects {
		mapped := ObjectSpec{Entry: Entry{ID: fmt.Sprintf("object_%d", objectIndex), AuditName: object.Name, ProgramName: object.Name}}
		for fieldIndex, field := range object.Fields {
			mapped.Fields = append(mapped.Fields, Entry{ID: fmt.Sprintf("field_%d_%d", objectIndex, fieldIndex), AuditName: field.Name, ProgramName: field.Name})
		}
		spec.Objects = append(spec.Objects, mapped)
	}
	for flowIndex, flow := range program.Flows {
		mapped := FlowSpec{Entry: Entry{ID: fmt.Sprintf("flow_%d", flowIndex), AuditName: flow.Name, ProgramName: flow.Name}}
		for inputIndex, input := range flow.Inputs {
			mapped.Inputs = append(mapped.Inputs, Entry{ID: fmt.Sprintf("input_%d_%d", flowIndex, inputIndex), AuditName: input.Name, ProgramName: input.Name})
		}
		for outputIndex, output := range flow.Outputs {
			mapped.Outputs = append(mapped.Outputs, Entry{ID: fmt.Sprintf("output_%d_%d", flowIndex, outputIndex), AuditName: output.Name, ProgramName: output.Name})
		}
		spec.Flows = append(spec.Flows, mapped)
	}
	return spec
}

func validateEntry(entry Entry, label string, ids map[string]bool, output *[]diagnostics.Diagnostic, file string) {
	if !stableIDPattern.MatchString(entry.ID) {
		*output = append(*output, diagnostic(file, "AAL2213", label+" has invalid stable id: "+entry.ID))
	}
	if ids[entry.ID] {
		*output = append(*output, diagnostic(file, "AAL2214", "binding id is duplicated: "+entry.ID))
	}
	ids[entry.ID] = true
	if entry.AuditName == "" || entry.ProgramName == "" {
		*output = append(*output, diagnostic(file, "AAL2215", label+" must have auditName and programName"))
	}
	if entry.ProgramName != "" && !programNamePattern.MatchString(entry.ProgramName) {
		*output = append(*output, diagnostic(file, "AAL2217", label+" has invalid programName: "+entry.ProgramName))
	}
}

func validateNamedEntries(entries []ObjectSpec, label string, output *[]diagnostics.Diagnostic, file string) {
	auditNames, programNames := map[string]bool{}, map[string]bool{}
	for _, entry := range entries {
		if auditNames[entry.AuditName] {
			*output = append(*output, diagnostic(file, "AAL2218", label+" auditName is duplicated: "+entry.AuditName))
		}
		if programNames[entry.ProgramName] {
			*output = append(*output, diagnostic(file, "AAL2219", label+" programName is duplicated: "+entry.ProgramName))
		}
		auditNames[entry.AuditName], programNames[entry.ProgramName] = true, true
	}
}

func validateNamedFlows(entries []FlowSpec, output *[]diagnostics.Diagnostic, file string) {
	auditNames, programNames := map[string]bool{}, map[string]bool{}
	for _, entry := range entries {
		if auditNames[entry.AuditName] {
			*output = append(*output, diagnostic(file, "AAL2218", "flow auditName is duplicated: "+entry.AuditName))
		}
		if programNames[entry.ProgramName] {
			*output = append(*output, diagnostic(file, "AAL2219", "flow programName is duplicated: "+entry.ProgramName))
		}
		auditNames[entry.AuditName], programNames[entry.ProgramName] = true, true
	}
}

func validateUniqueMemberNames(entries []Entry, label string, output *[]diagnostics.Diagnostic, file string) {
	names, programs := map[string]bool{}, map[string]bool{}
	for _, entry := range entries {
		if names[entry.AuditName] {
			*output = append(*output, diagnostic(file, "AAL2220", label+" auditName is duplicated: "+entry.AuditName))
		}
		if programs[entry.ProgramName] {
			*output = append(*output, diagnostic(file, "AAL2221", label+" programName is duplicated: "+entry.ProgramName))
		}
		names[entry.AuditName], programs[entry.ProgramName] = true, true
	}
}

func findObjectSpec(entries []ObjectSpec, name string) (ObjectSpec, bool) {
	for _, entry := range entries {
		if entry.AuditName == name {
			return entry, true
		}
	}
	return ObjectSpec{}, false
}

func findFlowSpec(entries []FlowSpec, name string) (FlowSpec, bool) {
	for _, entry := range entries {
		if entry.AuditName == name {
			return entry, true
		}
	}
	return FlowSpec{}, false
}

func findEntry(entries []Entry, name string) (Entry, bool) {
	for _, entry := range entries {
		if entry.AuditName == name {
			return entry, true
		}
	}
	return Entry{}, false
}

func find(objects []ast.Object, name string) (ast.Object, bool) {
	for _, object := range objects {
		if object.Name == name {
			return object, true
		}
	}
	return ast.Object{}, false
}

func hasFlow(program *ast.Program, name string) bool {
	for _, flow := range program.Flows {
		if flow.Name == name {
			return true
		}
	}
	return false
}

func findFlowInput(flow ast.Flow, name string) (ast.Field, bool) {
	for _, input := range flow.Inputs {
		if input.Name == name {
			return input, true
		}
	}
	return ast.Field{}, false
}

func hasObjectField(object ast.Object, name string) bool {
	for _, field := range object.Fields {
		if field.Name == name {
			return true
		}
	}
	return false
}

func findFlowOutput(flow ast.Flow, name string) (ast.Output, bool) {
	for _, output := range flow.Outputs {
		if output.Name == name {
			return output, true
		}
	}
	return ast.Output{}, false
}

func diagnostic(file, code, message string) diagnostics.Diagnostic {
	return diagnostics.Diagnostic{Severity: diagnostics.Error, Code: code, Message: message, File: file, Line: 1, Column: 1}
}

func at(location ast.Location, code, message string) diagnostics.Diagnostic {
	return diagnostics.Diagnostic{Severity: diagnostics.Error, Code: code, Message: message, File: location.File, Line: location.Line, Column: location.Column}
}
